### DISCUSSION TITLE
Upgrade from 2.2.5 to 2.3.0 using  apache/airflow docker image - ODBC issue

### DISCUSSION THREAD

**Discussion by thaysa** [apache/airflow#23561] [General] [answered]
Issue: **Tasks that use pyodbc to connect to postgres are running fine in 2.2.5 but they are throwing exception in 2.3.0.**

I'm investigating what has been changed on the official 2.3.0 airflow image (https://hub.docker.com/r/apache/airflow) in order to identify the root cause.

1. The 2.2.5 git version starts with **release** and the 2.3.0 starts with **dev**:  _Is dev in git version expected ?_
2. postgresql-client is not a dependency  in RUNTIME_APT_DEPS anymore
3. Is this pyodbc  issue a known issue in version 2.3.0? (customization: note that odbc-postgresql is used be installed on the top of the official image but applying the usual customization ... the pyodbc connection didn't work)
4. There is also an issue fetching remote logs in version 2.3.0 for tasks executed in previous airflow version. There was no fetching log issue during the migration from 2.2.4 to 2.2.5 and other old migrations. 
5. I would like to know if there are more people facing the same issue.
    
###  2.2.5 docker image :
**Git Version**: .**release**:2.2.5+7c3513903462511212df742263a49ec1f729ec0e

`docker pull apache/airflow:2.2.5-python3.9`

` docker inspect <image id> | grep RUNTIME_APT_DEPS`

"RUNTIME_APT_DEPS=       apt-transport-https        apt-utils        ca-certificates        curl        dumb-init        freetds-bin        gosu        krb5-user        ldap-utils        **libffi6**        libldap-2.4-2        libsasl2-2        libsasl2-modules        libssl1.1        locales         lsb-release        netcat        openssh-client        **postgresql-client**        **python2**        rsync        sasl2-bin        sqlite3        sudo        unixodbc",
                "ADDITIONAL_RUNTIME_APT_DEPS=",

/opt/airflow$ **ls /usr/lib/x86_64-linux-gnu/odbc/**
```
libesoobS.so  libnn.so            libodbcdrvcfg2S.so  libodbcmyS.so  libodbcpsqlS.so  liboplodbcS.so  libsapdbS.so
libmimerS.so  libodbcdrvcfg1S.so  libodbcminiS.so     libodbcnnS.so  libodbctxtS.so   liboraodbcS.so  libtdsS.so
```

###  2.3.0 docker image  :
**Git Version**: .**dev**0+6b772f68db4f80b01ede836ddd47437d4aa1b43d.**dirty**

docker pull **--platform=linux/amd64** apache/airflow:2.3.0-python3.9

`docker inspect <image id> | grep RUNTIME_APT_DEPS`

 "RUNTIME_APT_DEPS=       apt-transport-https        apt-utils        ca-certificates        curl        dumb-init        freetds-bin        gosu        krb5-user        ldap-utils        libldap-2.4-2        libsasl2-2        libsasl2-modules        libssl1.1        locales         lsb-release        netcat        openssh-client        rsync        sasl2-bin        sqlite3        sudo        unixodbc",
                "ADDITIONAL_RUNTIME_APT_DEPS=",

/opt/airflow$ **ls /usr/lib/x86_64-linux-gnu/odbc/**

`ls: cannot access '/usr/lib/x86_64-linux-gnu/odbc/': No such file or directory`

**Comment by potiuk** [comment, marked as answer]
First of all - Thanks for analysing the differences. That's good and helpful information.

Secondly - there is no guarantee that all the variables will remain the same between different versions of images. The images released by Airflow are just convenience packaging that you **might** use but we do not provide guarantees it will keep 100% compatibilty. You can read more about it here: https://airflow.apache.org/docs/docker-stack/changelog.html

So expectations that things will not change in the image are a bit too much.

You can always build your image following Image customisation here - if you want some specific dependencies:
https://airflow.apache.org/docs/docker-stack/build.html#customizing-the-image

However we try to keep compatibilty anyway. So maybe we can find out what happened and disect your analysis one by one:

> The 2.2.5 git version starts with release and the 2.3.0 starts with dev: Is dev in git version expected ?

Yeah. Airlfow 2.3.0 release process had some last-minute changes in release notes that resulted in the dev0 versoin and .dirty. We already know it and we fix our release processes so that it can be avoided (@jedcunningham @ephraimbuddy :) - users ARE paying attention :D). But this is harmless in this case.

>  postgresql-client is not a dependency in RUNTIME_APT_DEPS anymore

This is not a problem - its' been added differently, there is. a separate flag and separate step that adds postgresql client - see this PR: https://github.com/apache/airflow/pull/21631 - it's no longer of RUNTIME_APT_DEPS but it is installed anyway

> Is this pyodbc issue a known issue in version 2.3.0? (customization: note that odbc-postgresql is used be installed on the top of the official image but applying the usual customization ... the pyodbc connection didn't work)

What issue? 

From what I see, the odbc libraries are there but not in `odbc` subdirectory. It's possible that diffferent dependency version of the odbc library change the place of installation - but this is nothing Airflow can guarantee or "stick to" - you should adapt your configuration to the version of odbc installed:

```
airflow@41d1e80f5b8b:/opt/airflow$ find  /usr/lib/x86_64-linux-gnu/ -name '*odbc*'
/usr/lib/x86_64-linux-gnu/libodbcinst.so.2
/usr/lib/x86_64-linux-gnu/libodbccr.so.2.0.0
/usr/lib/x86_64-linux-gnu/libodbc.so.2
/usr/lib/x86_64-linux-gnu/libodbc.so.2.0.0
/usr/lib/x86_64-linux-gnu/libodbccr.so.1
/usr/lib/x86_64-linux-gnu/libodbc.so.1
/usr/lib/x86_64-linux-gnu/libodbcinst.so.2.0.0
/usr/lib/x86_64-linux-gnu/libodbccr.so.2
/usr/lib/x86_64-linux-gnu/libodbcinst.so.1
```

> There is also an issue fetching remote logs in version 2.3.0 for tasks executed in previous airflow version. There was no fetching log issue during the migration from 2.2.4 to 2.2.5 and other old migrations.

I think you need to open a separate issue with all details for that. It has nothing to do with any of the things mentioned here.

> I would like to know if there are more people facing the same issue.

Just search in issues/devlist. Airflow is a fully transparent and open-source project - everything here is in public - so you have precisely the same tools and possibilities to check if there are similar issues as anyone else - including maintainers. There is nothing "more" anyone in the project has capabilities of - just search.


Summarizing it - nothing in the question is "wrong" - all seems as expected (except the unknown fetch logs problem you mentioned - but this is another issue).

**Comment by thaysa** [comment]
@potiuk  thank you for your prompt response and all the details that you shared.
